cleanup/cleanup.sh

```diff
@@ -1,3 +1,111 @@
-#!/bin/sh
+#!/usr/bin/env bash
 
-echo "Plugin called."
+IMAGE_NAME=$1
+
+set -e -o pipefail -u
+
+GIT_COMMIT_LIMIT=${KUBECTL_PLUGINS_LOCAL_FLAG_GIT_COMMIT_LIMIT}
+GIT_REPO_PATH=${KUBECTL_PLUGINS_LOCAL_FLAG_GIT_REPO_PATH}
+FORCE=${KUBECTL_PLUGINS_LOCAL_FLAG_FORCE}
+KEEP=${KUBECTL_PLUGINS_LOCAL_FLAG_KEEP}
+NAMESPACE=${KUBECTL_PLUGINS_CURRENT_NAMESPACE}
+oc="${KUBECTL_PLUGINS_CALLER} -n ${NAMESPACE}"
+
+get_commit_list() {
+    # all commits except HEAD in the Git repository
+    git -C ${GIT_REPO_PATH} rev-list --max-count ${GIT_COMMIT_LIMIT} --author-date-order HEAD~
+}
+
+get_imagestreamtags() {
+    local imagestream_path='{range .status.tags[*]}{.tag}{"\n"}{end}'
+
+    ${oc} get imagestream ${IMAGE_NAME} -o jsonpath="${imagestream_path}" | sort --unique
+}
+
+get_active_imagestreamtags() {
+    local dc_rc_sfs_path='{range .items[*]}{.spec.template.spec.containers[*].image}{"\n"}{end}'
+    local pod_cont_path='{range .items[*]}{.spec.containers[*].image}{"\n"}{end}'
+    local pod_init_path='{range .items[*]}{.spec.initContainers[*].image}{"\n"}{end}'
+    local cron_path='{range .items[*]}{.spec.jobTemplate.spec.template.spec.containers[*].image}{"\n"}{end}'
+
+    local images_in_dc_rc_sset=$(${oc} get dc,rc,statefulset -o jsonpath="${dc_rc_sfs_path}")
```

Comment:
This will mask errors that happen inside `$()`, rendering `set -e` useless. To fix, use this:

    local images_in_dc_rc_sset
    images_in_dc_rc_sset=$(${oc} get dc,rc,statefulset -o jsonpath="${dc_rc_sfs_path}")

Here and later in the code

Reply:
Trying out the resulting code, this leads to exactly the same behavior: For example, a failing `git` command ("not a git repository"), though failing with an exit code of 128, will not abort the script:
```bash
$ oc plugin cleanup Foo -p ~
Comparing commits from /home/peter in namespace default
fatal: not a git repository (or any of the parent directories): .git
Error from server (NotFound): imagestreams.image.openshift.io "Foo" not found
No image tags found for deletion.
```

Reply:
Try the following script to reproduce the behaviour:
```sh
#!/bin/bash
set -e -u -o pipefail

main() {
  local foo
  foo=$(foo)
  echo "\$foo: $foo"
}

main
```

Reply:
Interesting ... :thinking:

Reply:
If we have another hierarchy level error detection fails again, e.g.

```bash
#!/usr/bin/env bash
set -e -o pipefail -u

do_git_status() {
    cd ~ && git status
}

get_deletion_candidates() {
    local foo
    foo=$(do_git_status)
    echo $foo
}

main() {
    local candidates
    candidates=($(get_deletion_candidates))
    echo "${candidates[@]}"
}

main
```
Can we get this under control, too?

Reply:
Ugh, this seems to be a hot mess of differences between bash versions and variants; the only way to _reliably_ exit when something blows up is to `set -e` inside **every function**